### PR TITLE
SR-56 Create Scheduled Inquiries (backend)

### PR DESCRIPTION
**Pull request by jakethrasher**
# Task
* Added FastAPI path operations for creating and getting all scheduled inquiries
* Updated the relationship between Inquiries and ScheduledInquiry to have `ondelete=CASCADE`. We don't want scheduled inquiries with references to inquiries that don't exist. This required an alembic migration.
* I refactored the test util in `test/utils/theme.py` to be reusable. This required small changes to code where the previous version was used.
* Added the service layer for scheduled inquiries
* Wrote tests for service and api layers

* Important note, this PR moves rank assignment to the backend when scheduled inquiries are created (see `app/services/scheduled_inquiries.py` `create` method).

[Link to Ticket](https://incredihire-academy.atlassian.net/browse/SR-56)

# Author Tasks

## Acceptance Criteria
Please include the acceptance criteria from the ticket here or select N/A.
- [x] N/A

### CODE REVIEW

File: backend/app/api/routes/scheduled_inquiries.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+)
+from app.services import (
+    scheduled_inquiries as scheduled_inquiries_service,
+)
+
+router = APIRouter()
+
+
+@router.post("/", response_model=ScheduledInquiryPublic)
+def create_scheduled_inquiry(
```
Comment (Eugene-Polonsky) [inline backend/app/api/routes/scheduled_inquiries.py:17]: I wonder if conceptually this isn't as much of a create operation as a "schedule an inquiry to be presented." I.e. we're not creating a scheduled inquiry, we're *scheduling* an inquiry. I get from the DB perspective it's an entry create, but it sounds a little weird from the overall conceptual perspective. Don't have a strong opinion either way, so feel free to ignore/resolve if you disagree.

File: backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    db_inquiry_2 = inquiries_service.create_inquiry(session=db, inquiry_in=inquiry_in_2)
+
+    yield db_inquiry_1, db_inquiry_2
+
+    # Teardown
+    db.delete(db_inquiry_1)
+    db.delete(db_inquiry_2)
+
+
+def test_get_scheduled_inquiries_when_scheduled_inquiries_exist_should_return_all(
```
Comment (Eugene-Polonsky) [inline backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py:31]: just wondering, should there be a case with an empty scheduled inquiries list?

File: backend/app/tests/api/routes/theme/test_post_theme_api.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 from fastapi.testclient import TestClient
 
 from app.models.theme import MAX_NAME_LENGTH, MIN_NAME_LENGTH
```
Comment (Eugene-Polonsky) [inline backend/app/tests/api/routes/theme/test_post_theme_api.py:6]: so this whole thing seems unrelated to the creating scheduled inquiries. Should this be a separate PR?

File: backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    db_scheduled_inquiry = db.get(ScheduledInquiry, scheduled_inquiry.id)
+
+    if not db_scheduled_inquiry:
+        raise Exception()
+
+    assert db_scheduled_inquiry.rank == expected_rank
+
+
+def test_create_when_scheduled_inquiries_exist_should_assign_correct_rank(
```
Comment (Eugene-Polonsky) [inline backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py:36]: I'd be more specific. e.g. 

test_create_when_one_scheduled_inquiry_exists_should_assign_rank_two

Comment (Eugene-Polonsky) [reply backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py:36]: also, and not sure if this is a JS convention, but at least in C#/C++ we camelCased it, e.g.

create_whenOneScheduledInquiryExists_shouldAssignRankTwo

seems a bit more readable. but either way.

File: backend/app/api/routes/scheduled_inquiries.py (hunk 1 above)
Comment (jakethrasher) [reply backend/app/api/routes/scheduled_inquiries.py:17]: Yeah, I think that is a valid point. I had some internal conflict about creating separate modules for `scheduled_inquiries`. Would it make more sense to add endpoints for operations on scheduled inquiries to `api/routes/inquiries.py` instead, and update the signature to something like:
```py
@router.post("/scheduledinquiries", response_model=ScheduledInquiryPublic)
def schedule_inquiry ()
...
```
In this case, there would be a POST to `/inquiries/scheduled_inquiries` which might be weird. I'm honestly not sure whats best.

Comment (jakethrasher) [reply backend/app/api/routes/scheduled_inquiries.py:17]: I could also just keep the separation between `inquiries` and `scheduled_inquiries`, and update the name to something more that makes more sense conceptually.

File: backend/app/tests/api/routes/theme/test_post_theme_api.py (hunk 3 above)
Comment (jakethrasher) [reply backend/app/tests/api/routes/theme/test_post_theme_api.py:6]: This was related to a small refactor to avoid duplication.

File: backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py (hunk 4 above)
Comment (jakethrasher) [reply backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py:36]: Yeah it looks god awful. Maybe we should go with camel case.

File: backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py (hunk 2 above)
Comment (jakethrasher) [reply backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py:31]: Caught me being lazy

File: backend/app/api/routes/scheduled_inquiries.py (hunk 1 above)
Comment (mcverter) [reply backend/app/api/routes/scheduled_inquiries.py:17]: I like this solution more
/inquiries/{inquiry_id}/schedule
because you have the same base artifact, the Inquiry, and you are just adding an attribute to it

Comment (Eugene-Polonsky) [reply backend/app/api/routes/scheduled_inquiries.py:17]: yeah, I like that too.

File: backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    client: TestClient,
+    superuser_token_headers: dict[str, str],
+    db: Session,
+) -> None:
+    # Setup
+    inquiry_in_1 = InquiryCreate(text="Test inquiry 1.")
+    inquiry_in_2 = InquiryCreate(text="Test inquiry 2.")
+    db_inquiry_1 = inquiries_service.create_inquiry(session=db, inquiry_in=inquiry_in_1)
```
Comment (ilhanbae) [inline backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py:20]: Should we avoid calling other services when testing a service? I had a similar doubt when I was writing test for `get_inquiries` service, and I ended up using fixture that directly creates an inquiry in the table.

File: backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from app.services import inquiries as inquiries_service
+from app.services import scheduled_inquiries as scheduled_inquiries_service
+
+
+@pytest.fixture
+def db_inquiry(db: Session) -> Generator[Inquiry]:
+    inquiry_in = InquiryCreate(text="I am test inquiry.")
+    db_inquiry = inquiries_service.create_inquiry(session=db, inquiry_in=inquiry_in)
+    yield db_inquiry
+    db.delete(db_inquiry)
```
Comment (ilhanbae) [inline backend/app/tests/services/scheduled_inquiries/test_create_scheduled_inquiry.py:16]: nice use of `db.delete` in fixture!

File: backend/app/tests/api/routes/scheduled_inquiry/test_add_to_schedule_api.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
+from fastapi.testclient import TestClient
+from sqlmodel import Session
+
+from app.models import InquiryCreate
+from app.services import inquiries as inquiries_service
+from app.tests.utils.api_post import api_post
+
+route_prefix = "scheduledinquiries"
+
+
```
Comment (ilhanbae) [inline backend/app/tests/api/routes/scheduled_inquiry/test_add_to_schedule_api.py:10]: Is there a DB clean up for after running API layer tests?

There's a cleanup fixture inside `confest.py` that we've been using to specify tables we wanted to cleanup after  running each test:
```
# Clear specified tables after running each test function
@pytest.fixture(scope="function", autouse=True)
def clear_tables_after_tests(db: Session) -> Generator[None, None, None]:
    yield
    tables_to_clear = [Inquiry]
    for table in tables_to_clear:
        statement = select(table)
        results = db.exec(statement).all()
        for record in results:
            db.delete(record)
    db.commit()
```

Comment (jakethrasher) [reply backend/app/tests/api/routes/scheduled_inquiry/test_add_to_schedule_api.py:10]: Thats part of the reason why I added the ondelete=CASCADE relationship between scheduled inquiries and inquiries. When this fixture runs it deletes the inquiries and scheduled inquiries by cascade. I'm not sure I love this fixture implementation. When you're dealing with relationships order matters when your deleting from the db. Since its doing a linear iteration here you have to make sure everything is in the right order for it to work. I'm not sure how well this will scale

File: backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py (hunk 5 above)
Comment (ilhanbae) [reply backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py:20]: Oh this is API test, but similar concern of isolation of unit test.

File: backend/app/tests/api/routes/scheduled_inquiry/test_add_to_schedule_api.py (hunk 7 above)
Comment (ilhanbae) [reply backend/app/tests/api/routes/scheduled_inquiry/test_add_to_schedule_api.py:10]: Got it. Yeah, I think this fixture function is not useful to scale. Ideally we want to automatically clear the tables that changed from each test. We could revisit this in future PR.

File: backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py (hunk 5 above)
Comment (jakethrasher) [reply backend/app/tests/api/routes/scheduled_inquiry/test_get_scheduled_inquiries_api.py:20]: That makes sense. I'll fix this

**Review by ilhanbae — DISMISSED** [review]

**Review by mcverter — DISMISSED** [review]

**Review by mcverter — APPROVED** [review]

**Review by ilhanbae — APPROVED** [review]
